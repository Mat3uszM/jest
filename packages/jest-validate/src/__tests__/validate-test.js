/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

const validate = require('../validate');
const defaultConfig = require('jest-config').defaults;
const {validConfig, deprecatedConfig} = require('jest-config');

test('validates default config', () => {
  expect(validate(defaultConfig, validConfig)).toBe(true);
});

[
  [{automock: []}, 'Boolean'],
  [{coverageReporters: {}}, 'Array'],
  [{preset: 1337}, 'String'],
  [{haste: 42}, 'Object'],
].forEach(([config, type]) => {
  test(`pretty prints valid config for ${type}`, () => {
    expect(() => validate(config, validConfig)).toThrowErrorMatchingSnapshot();
  });
});

test('omits null and undefined config values', () => {
  const config = {
    haste: undefined,
    preset: null,
  };
  expect(validate(config, validConfig)).toBe(true);
});

test('displays warning for unknown config options', () => {
  const config = {unknown: {}};
  const warn = console.warn;
  console.warn = jest.fn();

  validate(config, validConfig);

  expect(console.warn.mock.calls[0][0]).toMatchSnapshot();
  console.warn = warn;
});

test('displays warning for deprecated config options', () => {
  const config = {scriptPreprocessor: 'test'};
  const warn = console.warn;
  console.warn = jest.fn();

  validate(config, validConfig, deprecatedConfig);

  expect(console.warn.mock.calls[0][0]).toMatchSnapshot();
  console.warn = warn;
});

test('works with custom warnings', () => {
  const config = {unknown: 'string'};
  const validConfig = {test: [1, 2]};
  const warn = console.warn;
  const options = {
    footer: '\n\n  custom footer',
    namespace: 'My Custom',
  };
  console.warn = jest.fn();

  validate(config, validConfig, {}, options);

  expect(console.warn.mock.calls[0][0]).toMatchSnapshot();
  console.warn = warn;
});

test('works with custom errors', () => {
  const config = {test: 'string'};
  const validConfig = {test: [1, 2]};
  const options = {
    footer: '\n\n  custom footer',
    namespace: 'My Custom',
  };

  expect(() => validate(config, validConfig, {}, options))
    .toThrowErrorMatchingSnapshot();
});
