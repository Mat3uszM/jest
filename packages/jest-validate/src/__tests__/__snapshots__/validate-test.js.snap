exports[`test displays warning for deprecated config options 1`] = `
"[33m[1m[1m‚óè [1mJest Deprecation Warning:[22m

  Option [1mscriptPreprocessor[22m was replaced by [1mtransform[22m, which support multiple preprocessors.

  Jest now treats your current configuration as:
  {
    [1m\"transform\"[22m: [1m{\".*\": \"test\"}[22m
  }

  Please update your configuration.

  [1mConfiguration documentation:[22m
  https://facebook.github.io/jest/docs/configuration.html
[39m"
`;

exports[`test displays warning for unknown config options 1`] = `
"[33m[1m[1m‚óè [1mJest Validation Warning:[22m

  Unknown option [1munknown[22m with value [1m{}[22m was found.
  This is either a typing error or a user mistake. Fixing it will remove this message.

  [1mConfiguration documentation:[22m
  https://facebook.github.io/jest/docs/configuration.html
[39m"
`;

exports[`test pretty prints valid config for Array 1`] = `
"[31m[1m[1m‚óè [1mJest Validation Error:[22m[39m[37m

  Option [1mcoverageReporters[22m must be of type:
    [32mArray[37m
  but instead received:
    [31mObject[37m

  Example:
  {
    [1m\"coverageReporters\"[22m: [1m[\"json\", \"text\", \"lcov\", \"clover\"][22m
  }[39m[31m

  [1mConfiguration documentation:[22m
  https://facebook.github.io/jest/docs/configuration.html
[39m"
`;

exports[`test pretty prints valid config for Boolean 1`] = `
"[31m[1m[1m‚óè [1mJest Validation Error:[22m[39m[37m

  Option [1mautomock[22m must be of type:
    [32mBoolean[37m
  but instead received:
    [31mArray[37m

  Example:
  {
    [1m\"automock\"[22m: [1mfalse[22m
  }[39m[31m

  [1mConfiguration documentation:[22m
  https://facebook.github.io/jest/docs/configuration.html
[39m"
`;

exports[`test pretty prints valid config for Object 1`] = `
"[31m[1m[1m‚óè [1mJest Validation Error:[22m[39m[37m

  Option [1mhaste[22m must be of type:
    [32mObject[37m
  but instead received:
    [31mNumber[37m

  Example:
  {
    [1m\"haste\"[22m: [1m{\"providesModuleNodeModules\": [\"react\", \"react-native\"]}[22m
  }[39m[31m

  [1mConfiguration documentation:[22m
  https://facebook.github.io/jest/docs/configuration.html
[39m"
`;

exports[`test pretty prints valid config for String 1`] = `
"[31m[1m[1m‚óè [1mJest Validation Error:[22m[39m[37m

  Option [1mpreset[22m must be of type:
    [32mString[37m
  but instead received:
    [31mNumber[37m

  Example:
  {
    [1m\"preset\"[22m: [1m\"react-native\"[22m
  }[39m[31m

  [1mConfiguration documentation:[22m
  https://facebook.github.io/jest/docs/configuration.html
[39m"
`;

exports[`test works with custom deprecations 1`] = `
"[33m[1m[1m‚óè [1mMy Custom Deprecation Warning:[22m

  Option [1mscriptPreprocessor[22m was replaced by [1mtransform[22m, which support multiple preprocessors.

  Jest now treats your current configuration as:
  {
    [1m\"transform\"[22m: [1m{\".*\": \"test\"}[22m
  }

  Please update your configuration.

  custom footer[39m"
`;

exports[`test works with custom errors 1`] = `
"[31m[1m[1m‚óè [1mMy Custom Validation Error:[22m[39m[37m

  Option [1mtest[22m must be of type:
    [32mArray[37m
  but instead received:
    [31mString[37m

  Example:
  {
    [1m\"test\"[22m: [1m[1, 2][22m
  }[39m[31m

  custom footer[39m"
`;

exports[`test works with custom warnings 1`] = `
"[33m[1m[1m‚óè [1mMy Custom Validation Warning:[22m

  Unknown option [1munknown[22m with value [1m\"string\"[22m was found.
  This is either a typing error or a user mistake. Fixing it will remove this message.

  custom footer[39m"
`;
